### ISSUE TITLE
Refactor how submission file revisions are tracked and displayed 

### ISSUE THREAD

**Issue by NateWr** [pkp/pkp-lib#6079]
The current handling of submission file revisions is some of our oldest code and can be a challenge to work with. This issue is for discussion about how to refactor this.

_This issue is not about the file storage mechanism itself, such as moving to a third-party library like [Flysystem](https://flysystem.thephpleague.com/v1/docs/). Decisions about that may impact this, though._

### The good parts
- The revision system ensures the full history of a file can be tracked for auditing purposes.
- The activity log provides an accessible record of this history.
- Access to files -- and particular versions of those files -- can be controlled in cases, like reviewers, where only specific revisions should be accessible.

### The bad parts
- The `fileId` and `revision` must be combined to provide a unique ID for each file. This is not efficient database design and requires some hoop-jumping in the code (especially the file access policies).
- Each time a file is promoted from one file stage to another, its history is not displayed in the UI. It is not possible (in the UI) to see the full lifecycle of a file through the editorial process.
- Multilingual data, such as name and other metadata, is not unique to a revision. Modifications to this data change it for old revisions as well.

### Suggestions
- A third-party system that implements a form of version control like git but without additional server dependencies. So far we have not found any candidates.
- A refactor of our existing system which requires us to handle the versioning.

**Comment by NateWr** [comment]
Here is an old mockup of a UI approach to showing file revisions. Notice how the file history shows the file's movement across all file stages.

![file-revisions](https://user-images.githubusercontent.com/2306629/87032758-fecc7700-c1dc-11ea-8a06-9147e87f50df.png)

**Comment by NateWr** [comment]
Here's a first draft idea on how we can preserve a lot of what we have. Two key changes:

1. The `fileId` stays with files throughout the entire lifecycle. A submission file promoted through review, copyediting and production would always the same ID. This would replace the `source_*` fields. Promoting a file from one stage to another would create a new revision of that file.
2. File uniqueness is tracked through a `uuid` of some sort (a hash, something from the file management tool, or an autoincrement key in the db).

It helps me to think through this from the perspective of our API, since that gives me an overview of what kinds of reading/writing operations are needed. The endpoints could look like this:

| Method | Endpoint | Description |
| --- |--- | --- |
| GET | `/submissions/1/files` | All files the user has access to, returning the latest revision they can access. |
| GET | `/submissions/1/files/1` | Latest revision of file 1 the user has access to. |
| GET | `/submissions/1/files/1/revisions` | All revisions of file 1 the user has access to. |
| GET | `/submissions/1/files/1/revisions/1` | Revision 1 of file 1 if the user has access. |
| POST | `/submissions/1/files` | Create a new file. Must provide the file stage and genre. |
| POST | `/submissions/1/files/1` | Create a new revision of the file based on the latest revision. To prevent branching, it should not be possible for someone to create a revision of a file in an old stage when it has been promoted to a later stage. So someone with access to copyediting would not be able to create a revision of a file that had one or more revisions in the production stage. |
| PUT | `/submissions/1/files/1` | Modify metadata of file 1 without creating a new revision. Stage can not be modified. |
| PUT | `/submissions/1/files/1/revisions/1` | Modify metadata of file 1 revision 1 without creating a new revision. Stage can not be modified. |
| DELETE | `/submissions/1/files/1` | Delete a file and all revisions. |
| DELETE `/submissions/1/files/1` | Delete a file and all revisions. |

What we don't get this way is support for revision branches. I don't _think_ we are using branches currently. But we may want to do so in the future for things like a document-centric workflow. One way to support this would be to create a new `fileId` for each branch. This would work for disposable branches like a review assignment with comments. But not if we want support for merging back from branches in some way.

**Comment by asmecher** [comment]
@NateWr, where you tend to think in API, I tend to think in database. Maybe that's revealing...

An additional downside that would be good to solve: unnecessary copies of duplicate files (e.g. when promoting a file from one stage to another, the physical file is duplicated).

Here's how I'd propose adjusting our current database structure...

**Needs Doing**
- The `submission_files` table gets a foreign key to `files` (indicating the physical file) in place of the PHP function that currently builds a file path/name.
- The `submission_files` table gets a proper autonumber of its own (`submission_file_id` in place of the `file_id` / `revision` pairing). Convert `source_file_id` / `source_revision` so that they only refer to `submission_file_id` instead. Everything that contains a file_id / revision pairing for identifiers uses this new one instead.

**Maybes**
- Filesystem metadata (file size, file type, date uploaded) are *removed from the database*. The filesystem can provide these adequately and I don't think we'll ever need to e.g. sort a long list on this. Not a must-do but no real downsides.
- To resolve unnecessary duplication of files and support URLs, introduce a new `files` table that maps to files in the `files_dir` or remote URLs. The main purpose of this is to give a `file_id` autonumber that can be used in other tables for foreign keys to maps to binaries (local files or remote URLs). This table would not include anything about stage, access permissions, etc. It is never updated, only inserted to (file upload) or deleted from (deletion of the last use of the file). We could potentially use this for other kinds of files as well in the future e.g. public files if we want.

I do think we need to revisit the revision/source file ID concepts and figure out how we want our versioning to look, but I can imagine that spiralling quickly out of scope. These are all proposals that I think we could fit into 3.2 pretty handily.

**Comment by NateWr** [comment]
> An additional downside that would be good to solve: unnecessary copies of duplicate files

Ah, interesting. That opens up a lot of potential. I like where this is going.  A couple questions:

1. Would you keep the directory structure that's used to store files now? Or do you imagine a flat directory with unique filenames used as the foreign key in `files`?

2. Do you see each revision of a file appearing in `submission_files`? Or is there only one submission file entry (and associated metadata) with revisions tracked in a separate table? (I'm wondering about how we would identify past revisions and whether we need to track changes to metadata across revisions.)

3. What would the foreign key in `files` be to indicate the physical file? A unique hash, an incrementing id? At the moment it looks like submission files contain info about their submission, file stage, revision and date.

**Comment by asmecher** [comment]
> 1. Would you keep the directory structure that's used to store files now? Or do you imagine a flat directory with unique filenames used as the foreign key in files?

On migration, we would need to scan through `submission_files` and create an entry in `files` for each, containing the path and filename and thus associating it with an ID.

When new files are uploaded, the path and filename get assigned by the creating code.

Thereafter file names/paths can be whatever we want, and we don't even need consistency. But I still think something like `{$contextId}/submissions/{$submissionId}/{$someSyntheticFilename}` is sensible.

> 2. Do you see each revision of a file appearing in submission_files? Or is there only one submission file entry (and associated metadata) with revisions tracked in a separate table? (I'm wondering about how we would identify past revisions and whether we need to track changes to metadata across revisions.)

Yes, each revision would appear in `files` (representing the physical file that got uploaded) and `submission_files` (representing its stage etc).

> 3. What would the foreign key in files be to indicate the physical file? A unique hash, an incrementing id? At the moment it looks like submission files contain info about their submission, file stage, revision and date.

It would only be the file path/name, or alternately a URL, and the ID. I like the idea of including a hash but obviously we wouldn't want to verify that every time we touch the file. The workflow concerns live in `submission_files`, and if we don't want to bite it off yet, they can largely stay the same as they currently are there (with your above suggestions incorporated).

**Comment by NateWr** [comment]
That all makes good sense. I'm wondering about how we link rows in the `submission_files` table together across the editorial lifecycle. If a submission is uploaded during submission, passed through the many stages, how are these linked together? 

Since revisions are handled by a `fileId` foreign key, could the `submission_file_id` live with the file throughout the editorial lifecycle? In this case it's not an autoincrement id in the table. Maybe we need another non-unique id?

I'm thinking particularly of cases where I want to say "get me the latest instance of this file, limited by the stages that I have access to". In such cases we need to know which row in the table represents the latest revision of a file. And in this context we think of a file that existed in submission and review files as _the same file_.

**Comment by NateWr** [comment]
@asmecher I've got the `files`/`submission_files` tables set up as described. It's going well but I'm not sure how to track revisions to a file. 

A new row in `submission_files` is created whenever a new file is uploaded or an existing file is promoted to a new stage. However, when the user uploads a revision of the file (no change to metadata or `fileStage`), we create a new row in `files` and update `submission_files.fileId`.

At this point we've lost the revision history. The old `fileId` is no longer attached to anything.

If we create a new row in `submission_files`, we'll have a reference to both `fileId`s, but no way to identify which `submission_file` is the latest revision. So both files would appear in the files grid. The `source_submission_file_id` tells us where a particular row came from, but can't tell us which one is the latest without pulling all of them and looping through them one-by-one.

I think we need a one-to-many relationship between `submission_files` and `files`, so a separate table:

```
+---------+--------------------+
| file_id | submission_file_id |
+---------+--------------------+
|      30 | 1                  |
|      32 | 1                  |
+---------+--------------------+
```

We can then pull out the latest `file` for a `submission_file`, without losing a record of past revisions. What do you think?

(I'm still interested in the questions in my last comment, but I think that a call would be a better place to talk through that. And it's not work that will probably get done for 3.3.)

**Comment by asmecher** [comment]
We can get the single latest row from among many using a join like this:
```
SELECT * FROM ...
JOIN submission_files sf ON (your join conditions here)
LEFT JOIN submission_files newer_than_newest ON (sf.file_id < newer_than_newest.file_id)
WHERE ... AND newer_than_newest.file_id IS NULL;
```
The `JOIN` brings in the newest submission file, and the `LEFT JOIN` and `IS NULL` together assert that there is no newer entry.

However, whether that's a good idea is another question. How much of the data in a new revision's `submission_files` entry will be wasted duplicates?

**Comment by NateWr** [comment]
> How much of the data in a new revision's submission_files entry will be wasted duplicates?

It depends on how much we want to track metadata and genre changes across revisions. If a newly uploaded revision creates a new `submission_files` row, then we can track changes which might be important in cases where a revision has different licensing, caption or creator information -- though in the long-term I think we'll only offer this metadata during the galley/publication format stage. It is of limited value during the editorial workflow.

If we don't want to track these changes, then pretty much all of the data is wasted duplicates. If we think of a `submission_file` as one thing with two `file` revisions, it probably makes most sense to use a second table with a familiar join.

This makes more sense to me when we consider attaching an activity log and notes to files as well. It makes sense for these things to carry through across revisions.

**Comment by asmecher** [comment]
I'm OK in theory with tracking changes to e.g. captions, creators, etc. in the `submission_files` table by adding new entries there when they're revised. However, we're now contemplating two mechanisms for versioning, and by our current design that means a new submission revision will mean full copies of all this submission file content as well.

I would suggest that we consider these new tables invariant, i.e. created once and *never* updated. That way we would not need to duplicate them when creating a new submission version.

**Comment by NateWr** [comment]
This work is being done as part of https://github.com/pkp/pkp-lib/issues/6057.

**Comment by NateWr** [comment]
@asmecher I've found references to two file stages which don't appear to be used anymore: `SUBMISSION_FILE_FAIR_COPY` and `SUBMISSION_FILE_EDITOR`. There's an upgrade script that sets `SUBMISSION_FILE_FAIR_COPY` when converting files over from 2.x to 3.0. But it's no longer used.

Is it ok to remove these? Do we need them in the system for posterity or old data retrieval of any kind?

**Comment by asmecher** [comment]
Yes, please remove those. I generally try to leave decommissioned constants in place with a "deprecated" note but now that the 2.4.x upgrade path has been removed from the `master` codebase it's just clutter.

**Comment by NateWr** [comment]
This was addressed in #6057.
